arch/apple2/decoder.cc

```diff
@@ -101,11 +104,28 @@ class Apple2Decoder : public AbstractDecoder
 		if (readRaw24() != APPLE2_DATA_RECORD)
 			return;
 
+		// Sometimes there's a 1-bit gap between APPLE2_DATA_RECORD and
+		// the data itself.  This has been seen on real world disks
+		// such as the Apple II Operating System Kit from Apple2Online.
+		// However, I haven't seen it described in any of the various
+		// references.
+		//
+		// This extra '0' bit would not affect the real disk interface,
+		// as it was a '1' reaching the top bit of a shift register
+		// that triggered a byte to be available, but it affects the
+		// way the data is read here.
+		auto p = tell();
+		auto b = readRawBits(1);
+		if(b[0]) seek(p);
```

Comment:
You can't do seek() here. The flux decoder has internal state which gets reset when a seek happens (or, more accurately, _should_ be reset when a seek happens and frequently isn't). I had to do a major refactor a few weeks ago to remove spurious seeks which were corrupting the bit stream.

It's probably possible to do this by fiddling with the flux decoder's state, but it might be easier to just read recordLength*8+1 bits and discard the first if it's a zero. There should be a gap afterwards so the extra bit shouldn't do any harm.

Reply:
I see, thanks. I'll reformulate it.

Reply:
Do you feel any better about doing it like this?
```c++
                // Sometimes there's a 1-bit gap between APPLE2_DATA_RECORD and
                // the data itself.  This has been seen on real world disks
                // such as the Apple II Operating System Kit from Apple2Online.
                // However, I haven't seen it described in any of the various
                // references.
                //  
                // This extra '0' bit would not affect the real disk interface,
                // as it was a '1' reaching the top bit of a shift register
                // that triggered a byte to be available, but it affects the
                // way the data is read here.
                //
                // While the floppies tested only seemed to need this applied
                // to the first byte of the data record, applying it
                // consistently to all of them doesn't seem to hurt, and
                // simplifies the code.
                
                /* Read and decode data. */   
                
                auto readApple8 = [&]() {
                    auto result = readRaw8();
                    while(result & 0x80 == 0) {
                        auto b = readRawBits(1);
                        result = (result << 1) | b[0];
                    }
                    return result;
                };
 
                constexpr unsigned recordLength = APPLE2_ENCODED_SECTOR_LENGTH + 2;
                uint8_t bytes[recordLength];
                for(auto &byte : bytes) {
                    byte = readApple8();
                }
```

Reply:
Yes, that looks fine. Will merge.